### PR TITLE
feat: add optional git cloning

### PR DESCRIPTION
**Pull request by JonZeolla**
# Contributor Comments

This adds a git cloning feature, so with a simple environment-based configuration (and providing in the right ssh keys/configurations on disk, when necessary) you can clone repositories at runtime prior to performing the scans.  This is the second stage of foundation laying to create a supported scanning factory-based approach (the first step was 238ba4f569970e8e724de24c18cfbdf08846a14c).

TDD evidence is [here](https://github.com/SeisoLLC/easy_infra/actions/runs/4037821735/jobs/6941388564#step:11:19).

- [x] Finalize tests.  Blocked on https://github.com/SeisoLLC/easy_infra/pull/235; jq parsing issues are causing issues with testing this.

## Pull Request Checklist

Thank you for submitting a contribution to `easy_infra`!

In order to streamline the review of your contribution we ask that you review and comply with the below requirements:

- [x] Rebase your branch against the latest commit of the target branch
- [x] If you are adding a dependency, please explain how it was chosen
- [x] If manual testing is needed in order to validate the changes, provide a testing plan and the expected results
- [x] If there is an issue associated with your Pull Request, link the issue to the PR.

### CODE REVIEW

File: tests/test.py [hunk 1]
```diff
@@ -1123,6 +1134,27 @@ def run_ansible(*, image: str) -> None:
...
 
+    # Test the git clone feature
+    command = "scan_ansible"
+    LOG.debug("Testing scan_ansible against a repository that was cloned at runtime")
+    environment = {}
+    environment["VCS_DOMAIN"] = "github.com"
+    environment["CLONE_REPOSITORIES"] = "seisollc/easy_infra,seisollc/easy_infra"
+    environment["CLONE_PROTOCOL"] = "https"
+
+    # TODO: In the future, migrate this to a general test config
```
Comment (JonZeolla) [inline tests/test.py:1145]: This is purposefully left in for posterity.  If anyone strongly disagrees we can consider another way to note this

File: tests/test.py [hunk 2]
```diff
@@ -1150,6 +1170,34 @@ def run_ansible(*, image: str) -> None:
...
+
+    # Purposefully missing volumes= because we are using clone to do it
+    utils.opinionated_docker_run(
+        image=image,
+        command=command,
+        environment=environment,
+        expected_exit=0,
```
Comment (JonZeolla) [inline tests/test.py:1197]:
```suggestion
        expected_exit=50,
```

File: tests/test.py [hunk 3]
```diff
@@ -1150,6 +1170,34 @@ def run_ansible(*, image: str) -> None:
...
+    #
+    # Running two commands that include security scans purposefully; there used to be a bug
+    #
+    # Note that we can't just replace the cd with workdir because it will create the dir as root prior to hitting the ENTRYPOINT, and the container
+    # user won't have access to write in that directory when it tries to run _clone. This was fixed in buildkit in
+    # https://github.com/moby/buildkit/pull/1002 but docker-py doesn't support buildkit yet; see the very popular
+    # https://github.com/docker/docker-py/issues/2230 issue from January 2019
+    #
+    # The exit 230 ensures that, if the dir doesn't exist, it doesn't accidentally match the expected_exit of 1 below
+    command = '/bin/bash -c "cd /iac/seisollc/easy_infra/tests/ansible/tool/kics || exit 230 && scan_ansible && scan_ansible"'
```
Comment (JonZeolla) [inline tests/test.py:1183]:
```suggestion
    command = '/bin/bash -c "cd /iac/seisollc/easy_infra/tests/ansible/tool/kics || exit 230 && scan_ansible || scan_ansible"'
```

File: tests/test.py [hunk 4]
```diff
@@ -380,17 +380,37 @@ def run_terraform(*, image: str) -> None:
...
+    #
+    # Running two commands that include security scans purposefully; there used to be a bug
+    #
+    # Note that we can't just replace the cd with workdir because it will create the dir as root prior to hitting the ENTRYPOINT, and the container
+    # user won't have access to write in that directory when it tries to run _clone. This was fixed in buildkit in
+    # https://github.com/moby/buildkit/pull/1002 but docker-py doesn't support buildkit yet; see the very popular
+    # https://github.com/docker/docker-py/issues/2230 issue from January 2019
+    #
+    # The exit 230 ensures that, if the dir doesn't exist, it doesn't accidentally match the expected_exit of 1 below
+    command = '/bin/bash -c "cd /iac/seisollc/easy_infra/tests/terraform/general/invalid || exit 230 && scan_terraform && terraform validate"'
```
Comment (JonZeolla) [inline tests/test.py:393]:
```suggestion
    command = '/bin/bash -c "cd /iac/seisollc/easy_infra/tests/terraform/general/invalid || exit 230 && scan_terraform || terraform validate"'
```

**Review by JonZeolla — COMMENTED** [review]
Tweaks from mobile

File: tests/test.py [hunk 5]
```diff
@@ -380,17 +380,37 @@ def run_terraform(*, image: str) -> None:
     )
     num_tests_ran += 1
 
-    # Test learning mode on an invalid configuration
-    command = "terraform validate"
-    LOG.debug("Testing learning mode on an invalid configuration")
+    # Test learning mode on an invalid configuration, using the git clone feature, non-interactively
+    #
+    # Running two commands that include security scans purposefully; there used to be a bug
```
Comment (JonZeolla) [inline tests/test.py:385]:
```suggestion
```

File: tests/test.py [hunk 6]
```diff
@@ -380,17 +380,37 @@ def run_terraform(*, image: str) -> None:
...
+    #
+    # Note that we can't just replace the cd with workdir because it will create the dir as root prior to hitting the ENTRYPOINT, and the container
+    # user won't have access to write in that directory when it tries to run _clone. This was fixed in buildkit in
+    # https://github.com/moby/buildkit/pull/1002 but docker-py doesn't support buildkit yet; see the very popular
+    # https://github.com/docker/docker-py/issues/2230 issue from January 2019
+    #
+    # The exit 230 ensures that, if the dir doesn't exist, it doesn't accidentally match the expected_exit of 1 below
+    command = '/bin/bash -c "cd /iac/seisollc/easy_infra/tests/terraform/general/invalid || exit 230 && scan_terraform || terraform validate"'
```
Comment (JonZeolla) [inline tests/test.py:393]:
```suggestion
    command = '/bin/bash -c "cd /iac/seisollc/easy_infra/tests/terraform/general/invalid || exit 230 && terraform validate"'
```

**Review by JonZeolla — COMMENTED** [review]
More mobile tweaks

File: docs/Cloning/index.rst [hunk 7]
```diff
@@ -0,0 +1,30 @@
+*******
+Cloning
+*******
+
+``easy_infra`` allows you to clone git repositories at runtime, based on the configuration options specified below.
+
+Errors encountered during the cloning process are written within the directory that the repos are downloaded in (defaults to ``/iac``), and if any
```
Comment (JonZeolla) [inline docs/Cloning/index.rst:7]:
```suggestion
Errors encountered during the cloning process are written to ``/var/log/clone.log``, and if there are any errors they are written to ``/var/log/clone.err.log``.
```

File: docs/Cloning/index.rst [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+fatal errors are encountered, a log is written to which can then be centralized for analysis if `Logging <../Logging/index.html>`_ is properly
```
Comment (JonZeolla) [inline docs/Cloning/index.rst:8]:
```suggestion
If fatal errors are encountered, logged are written to which can that will centralized for analysis if `Logging <../Logging/index.html>`_ is properly
```

**Review by digeseiso — APPROVED** [review]

**Pull request merged**